broadinstitute/single_cell_portal_core
# React upload wizard required fields & validation (SCP-3740)

This adds the validations detailed in the [google doc](https://docs.google.com/document/d/1DEWrROU-zgovNaK9lR_SnshCf7tmCdBoz3TihRsgWCM/edit#heading=h.29woqp4kz24g).  

Name uniqueness and file extension errors are shown directly under the file chooser
![image](https://user-images.githubusercontent.com/2800795/135947733-9cd48b5c-bd10-47e6-8c95-4185550c24f4.png)
![image](https://user-images.githubusercontent.com/2800795/135947765-b166956b-0fc4-459a-83ea-479fed3749f4.png)
(note the latter screenshot is actually difficult to get, since for the most part you are prevented from even  selecting files without the approved extensions.  But for some weird reason the OS X file chooser lets you pick '.js' files no matter what types you pass in? 

Required fields are indicated with an asterisk next to the field label, and the save button displays a tooltip on hover/focus indicating any missing required fields (as well as repeating any file name / extension errors)
![image](https://user-images.githubusercontent.com/2800795/135950214-6fd87767-1dc1-4f18-be8f-3ea3ea9efb9f.png)


This also includes the start to what I hope will become good automated test coverage of the wizard frontend.  I've created more jira tickets for large areas needing additional automation coverage.

TO TEST:
  1. Create a new study (with react upload wizard feature flag on)
  2. attempt to save the empty raw counts file that appears
  3. confirm the validation errors appear
  3. fill out the raw counts form, selecting a valid file
  4. confirm you can save the raw counts file
  5. move to processed matrix step.
  6. Attempt to select the same file you uploaded for raw counts
  7. confirm a 'that file already exists in your study' message appears
  8. Move through the other steps, confirming that file validation and required field validation is present
